### ISSUE TITLE
Whitelisting of the MacFuse System Extension is not working

### ISSUE THREAD

**Issue by TK5-Tim** [osxfuse/osxfuse#879]
Hello everyone, 

We are looking to distribute MacFuse to our Users so they can use VeraCrypt properly on their Mac devices. To be able to use MacFuse the system extension has to be approved. Using a MDM profile with the Team Identifier and Identifier of the System Extension to whitelist it in the system. 
But unfortunately whitelisting does not have the effect it should and instead it still needs extra approval with admin right to allow the system extension. 

Do you know anything that would be possible for me to do to solve this problem? 

Thank you very much!

**Comment by ferdnyc** [comment]
I think the problem is that MDM whitelisting [is for System Extensions](https://developer.apple.com/documentation/systemextensions/implementing_drivers_system_extensions_and_kexts), which are different from kernel extensions (kexts) – there doesn't seem to be any whitelisting or automated management of those anymore, since Apple is clearly trying hard to phase them out entirely. 

So [filesystem kexts don't qualify](https://developer.apple.com/documentation/devicemanagement/systemextensions): "The allowed extension types are `DriverExtension`, `NetworkExtension`, and `EndpointSecurityExtension`." The admin approval, not to mention the subsequent reboot, have been made non-negotiable.

(Heck, for Silicon machines [it's even _worse_](https://developer.apple.com/documentation/apple-silicon/installing-a-custom-kernel-extension/):)

> ### Explain the Kext-Loading Process to Users
> 
> In macOS 11 and later, the loading of a kext requires the user to take actions outside of your installation package. When loading a kext, tell the user to perform the following one-time setup on Apple silicon:
> 
> 1. Reboot your Mac with Apple silicon into Recovery mode.
> 1. Set the security level to Reduced security.
> 1. Allow the loading of third-party kexts.
> 1. Reboot back to macOS.

I'm thinking of trying to convince GitHub to preinstall MacFuse on their osx-latest runner images for GitHub Actions, since there doesn't appear to be any hope of installing it unattended, much less _during_ a single job run without reboot, for any software that needs it when performing tests. (The reason [vasi/squashfuse](https://github.com/vasi/squashfuse/)'s Mac CI is currently failing.)

**Comment by TK5-Tim** [comment]
Thank you for your answer. Unfortunately we cannot use this workaround in our situation. 

Is there any roadmap to migrate from a kernel extension to a system extension or is that not feasible? 

Thank you very much once again for the information.

**Comment by bfleischer** [comment]
@TK5-Tim Currently, there is no system extension API for developing full-featured file systems. Until Apple makes such an API available, macFUSE needs to rely on the VFS kernel API.

**Comment by ferdnyc** [comment]
(Actually, @bfleischer, I guess based on the license terms, GitHub/Microsoft would require permission to bundle a preinstalled MacFuse kext with any of their MacOS runner images for GitHub Actions.

In the _very_ unlikely event that they could be convinced to do so, would you be open to granting them authorization? It's not even worth pursuing if the answer is "No". :grin: )

**Comment by mcajina74** [comment]
Using JamfPro, I was able to get past this issue using an "**Approved Kernel Extensions**" payload in a Configuration Profile. You have to check both boxes, "**Allow users to approve kernel extensions**" and "**Allow standard users to approve legacy kernel extensions (macOS 11 or later)**." This will allow a non-Admin user to approve the KEXT. Please see my screen shot below. I tested this in both macOS Big Sur and Monterey.

<img width="827" alt="macFUSE_AppvdKEXT" src="https://user-images.githubusercontent.com/11951639/171957164-89d7a2a2-67ea-4778-82b7-cc1807f5ded9.png">


The only weird issue I see with the installation is that two different warning popups come up asking you to go to System Preferences to approve the System Extension. A user can get confused, but either one should take them to the Security & Privacy PrefPane to click the "**Allow**" button. They will then need to restart.

<img width="372" alt="Screen Shot 2022-06-03 at 11 28 06 AM" src="https://user-images.githubusercontent.com/11951639/171958259-6bf20788-bab2-48bb-9c48-470617da8fef.png">
<img width="372" alt="Screen Shot 2022-06-03 at 11 28 10 AM" src="https://user-images.githubusercontent.com/11951639/171958278-d54d52be-6c34-42f9-b0be-e9b753a8f90b.png">
<img width="780" alt="Screen Shot 2022-06-03 at 11 29 22 AM" src="https://user-images.githubusercontent.com/11951639/171958540-fcebe5c4-2080-4cf1-bf3b-1d6246816838.png">

**Comment by ferdnyc** [comment]
Interesting, and encouraging, @mcajina74 !

> I tested this in both macOS Big Sur and Monterey.

Have you been able to test it on Silicon hardware, by any chance? There, I fear, is where this will unfortunately break down. Apple have made it very clear they consider installing kexts on Silicon Macs to be a security compromise only permissible when booted into the system's recovery environment. :disappointed: 

Still, even if it only works for Intel Macs, at least it's something! (Though the forced reboot means it's also still unusable when running in unattended/transient environments, like GitHub Actions CI jobs.)

**Comment by mcajina74** [comment]
@ferdnyc , I did not test on an M1 Mac. I apologize for not mentioning that I tested on Intel. Since I don't have an M1 Mac available for testing, I tend to forget they exist.

Yes, the restart is a bit annoying, but the System Extension warnings are also confusing to the end user. One says it's blocked, and the other says it's been updated when it's a fresh installation. Either one will take you to the Security & Privacy to "Allow" it, but perhaps we only need one to come up.

EDIT: I want to add that the System Extension (or Kernel Extension) warnings did not come up for computers running Catalina (Intel). I checked the Security & Privacy prefpane and there was no "Allow" button to click, and therefore no restart warning.

**Comment by kfattic** [comment]
I wanted to mention an issue I had with this as well... we use Mosyle as our MDM.  We also had to whitelist the Team ID as mentioned above in this thread, but it won't work until you do a special restart and rebuild the kernel cache as shown in the picture below.  You need to click on the Advanced Options.  Once this happens, you should be able to get your software working correctly.  In this case we are using mLogic Canister for LTO backup.  Editing this comment to note that this is for an Apple Silicon M1.

![CleanShot 2023-03-27 at 07 59 10](https://user-images.githubusercontent.com/6775519/227946203-d1b8a05c-bace-4b71-9fbb-9c67e93a5551.png)

**Comment by kfattic** [comment]
I wanted to mention an issue I had with this as well... we use Mosyle as our MDM.  We also had to whitelist the Team ID as mentioned above in this thread, but it won't work until you do a special restart and rebuild the kernel cache as shown in the picture below.  You need to click on the Advanced Options.  Once this happens, you should be able to get your software working correctly.  In this case we are using mLogic Canister for LTO backup.

![CleanShot 2023-03-27 at 07 59 10](https://user-images.githubusercontent.com/6775519/227946203-d1b8a05c-bace-4b71-9fbb-9c67e93a5551.png)

**Comment by kfattic** [comment]
I am providing another update as we were presented with an issue when updating from macOS 13.2.1 to macOS 13.3.1.  The message from the OS was that the System Extensions needed to be rebuilt and to go into System Settings to approve/allow that change.  Unfortunately, going into System Settings would not bring up the option to approve/allow it.  To be clear, this was for a Mac that already has the system and kernel extensions approved/allowed from the MDM.  It was already in Reduced Security (which is what happens by default when approving/allowing Kext's from an MDM) and also had both options in the bottom part of that window checked in Reduced Security (this is visible by booting into Recovery - https://support.apple.com/guide/mac-help/change-security-settings-startup-disk-a-mac-mchl768f7291/mac).  The only thing that worked to get this Apple Silicon M1 iMac to get the approve/allow message(s) to appear (which again, you would think were not needed as the approvals were already coming from the MDM... yet in order to rebuild the System Extensions, you need the prompt to appear) was to disable SIP.  I had to do two approve/allow options.  One was to rebuild the System Settings and then restart.  The next was to approve/allow macFUSE again along with another restart.  Everything worked as expected afterwards.

TL;DR - Disable SIP while troubleshooting to get approve/allow message to appear in System Settings to rebuild System Extensions.  I don't believe this should be necessary and have filed FB with Apple: FB12166580.

**Comment by ravenep** [comment]
Hi @kfattic thanks for providing your solution. I have a similar problem, but slightly different, and am hoping to get feedback here. If this is off topic, I am also happy to submit a new issue. 

I have a 14in 2021 Apple M1 Pro, running Ventura 13.6.2

After following the install tutorial, I got stuck with the error below. I did as you suggested and disabled the SIP. This did not seem to make a difference. I have now restarted my computer so many times it's becoming a bit silly, so thought I'd reach out for some help. Any suggestions welcome! thanks

<img width="305" alt="error" src="https://github.com/osxfuse/osxfuse/assets/12449573/423af680-1167-43bb-be28-2263dd13f92e">

**Comment by ferdnyc** [comment]
@ravenep 

Just to confirm, you've also booted into Recovery and configured the system for Reduced Security mode? (IOW, you've seen this dialog, or set equivalent options in your MDM if you're using one?):

![image](https://github.com/osxfuse/osxfuse/assets/538020/beb551cd-9e72-4ce3-be5a-08eb40db69a1)

**Comment by ravenep** [comment]
Thanks for the reply, yes I did disable. I also went to the next level and disabled SIP. I opened issue #988 with more details of the current status (ie still not working :))
